examples/multisig_plain/lib.rs

```diff
@@ -0,0 +1,837 @@
+    struct Confirmation {
+        /// The transaction that was confirmed.
+        #[ink(topic)]
+        transaction: TransactionId,
+        /// The owner that sent the confirmation.
+        #[ink(topic)]
+        from: AccountId,
+    }
+
+    /// Emitted when an owner revoked a confirmation.
+    #[ink(event)]
+    struct Revokation {
+        /// The transaction that was revoked.
+        #[ink(topic)]
+        transaction: TransactionId,
+        /// The owner that sent the revokation.
+        #[ink(topic)]
+        from: AccountId,
+    }
+
+    /// Emitted when an owner submits a transaction.
+    #[ink(event)]
+    struct Submission {
+        /// The transaction that was submitted.
+        #[ink(topic)]
+        transaction: TransactionId,
+    }
+
+    /// Emitted when a transaction was canceled.
+    #[ink(event)]
+    struct Cancelation {
+        /// The transaction that was canceled.
+        #[ink(topic)]
+        transaction: TransactionId,
+    }
+
+    /// Emitted when a transaction was executed.
+    #[ink(event)]
+    struct Execution {
+        /// The transaction that was executed.
+        #[ink(topic)]
+        transaction: TransactionId,
+        /// Indicates whether the transaction executed successfully.
+        #[ink(topic)]
+        result: Result<(), ()>,
+    }
+
+    /// Emitted when an owner is added to the wallet.
+    #[ink(event)]
+    struct OwnerAddition {
+        /// The owner that was added.
+        #[ink(topic)]
+        owner: AccountId,
+    }
+
+    /// Emitted when an owner is removed from the wallet.
+    #[ink(event)]
+    struct OwnerRemoval {
+        /// The owner that was removed.
+        #[ink(topic)]
+        owner: AccountId,
+    }
+
+    /// Emitted when the requirement changed.
+    #[ink(event)]
+    struct RequirementChange {
+        /// The new requirement value.
+        new_requirement: u32,
+    }
+
+    impl MultisigPlain {
+        /// The only constructor of the contract.
+        /// A list of owners must be supplied and a number of how many of them must
+        /// confirm a transaction. Duplicate owners are silently dropped.
+        #[ink(constructor)]
+        fn new(&mut self, owners: Vec<AccountId>, requirement: u32) {
+            for owner in &owners {
+                self.is_owner.insert(*owner, ());
+                self.owners.push(*owner);
+            }
+            self.ensure_requirement_is_valid(self.owners.len(), requirement);
+            assert!(self.is_owner.len() == self.owners.len());
+            self.requirement.set(requirement);
+        }
+
+        /// Add a new owner to the contract.
+        /// Panics if the owner already exists.
+        /// Only callable by the wallet itself.
+        #[ink(message)]
+        fn add_owner(&mut self, new_owner: AccountId) {
+            self.ensure_from_wallet();
+            self.ensure_no_owner(&new_owner);
+            self.ensure_requirement_is_valid(self.owners.len() + 1, *self.requirement);
```

Comment:
Btw.: I really like the defensive programming style used in this PR. :) Really good example of how contracts should be written.